google-cloud-firestore/samples/query_watch.rb

```diff
@@ -0,0 +1,199 @@
+# Copyright 2020 Google, Inc
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google/cloud/firestore"
+
+def listen_document project_id:, collection_path: "cities"
+  # project_id = "Your Google Cloud Project ID"
+  # collection_path = "cities"
+
+  firestore = Google::Cloud::Firestore.new project_id: project_id
+  # [START fs_listen_document]
+  doc_ref = firestore.col(collection_path).doc "SF"
+  snapshots = []
+
+  # Watch the document.
+  listener = doc_ref.listen do |snapshot|
+    # Process the snapshot.
+    puts "Received document snapshot: #{snapshot.document_id}"
+    snapshots << snapshot
+  end
+  # [END fs_listen_document]
+
+  sleep 1
+
+  # Create the document.
+  doc_ref.set(
+    name:       "San Francisco",
+    state:      "CA",
+    country:    "USA",
+    capital:    false,
+    population: 860_000
+  )
+  # Wait for the callback.
+  wait_until { snapshots.count == 1 }
+
+  # [START fs_detach_listener]
+  listener.stop
+  # [END fs_detach_listener]
+end
+
+def listen_changes project_id:, collection_path: "cities"
+  # project_id = "Your Google Cloud Project ID"
+  # collection_path = "cities"
+
+  firestore = Google::Cloud::Firestore.new project_id: project_id
+  # [START fs_listen_changes]
+  query = firestore.col(collection_path).where :state, :==, "CA"
+  snapshots = []
+
+  # Watch the collection query.
+  listener = query.listen do |snapshot|
+    puts "Callback received query snapshot."
+    puts "Current cities in California:"
+    snapshot.changes.each do |change|
+      # Process the snapshot.
+      if change.added?
+        puts "New city: #{change.doc.document_id}"
+      elsif change.modified?
+        puts "Modified city: #{change.doc.document_id}"
+      elsif change.removed?
+        puts "Removed city: #{change.doc.document_id}"
+        snapshots << snapshot
+      end
+    end
+  end
+  # [END fs_listen_changes]
+
+  mtv_doc = firestore.col(collection_path).doc("MTV")
+
+  sleep 1
+
+  # Create the document.
+  mtv_doc.set(
+    name:       "Mountain View",
+    state:      "CA",
+    country:    "USA",
+    capital:    false,
+    population: 80_000
+  )
+
+  sleep 1
```

Comment:
Can we block on the previous operation rather than sleeping? Even with this sleep, the order is not guaranteed.

Reply:
I suppose that if `snapshots` is something more complex, like a hash, we could wait for each type of change.

Reply:
Done.